File: IN/sample.cpp
// -- IMPORTS

#include "Engine/Core/Common.h"
#include "Engine/Core/Collections/Dictionary.h"
#include "Engine/Core/Log.h"

// -- CONSTANTS

constexpr int32
    MinimumPassCount = 0,
    MaximumPassCount = 5;

// -- TYPES

class Being
{
    // == PUBLIC

    public:

    // -- ATTRIBUTES

    String
        Name;
    int32
        Age;

    // -- CONSTRUCTORS

    Being(
        String name,
        int32 age
        )
    {
        Name = name;
        Age = age;
    }
};

// ~~

class Person:
    public Being
{
    // == PUBLIC

    public:

    // -- ATTRIBUTES

    double
        Weight,
        DogCount;

    // -- CONSTRUCTORS

    Person(
        String name,
        int32 age,
        double weight,
        double dogCount
        ):
        Being( name, age )
    {
        Weight = weight;
        DogCount = dogCount;
    }

    // -- INQUIRIES

    int32 GetAge(
        )
    {
        return Age;
    }

    // ~~

    int32 GetAgeOffset(
        int32 otherAge
        )
    {
        return otherAge - Age;
    }

    // ~~

    String GetHelloMessage(
        )
    {
        return
            String::Format(
                TEXT( "Hello, my name is {0}, I'm {1} years old and I weight {2} kilograms." ),
                Name,
                Age,
                Weight
                );
    }

    // -- OPERATIONS

    void SetAge(
        int32 age
        )
    {
        Age = age;
    }

    // ~~

    void SetFakeAge(
        int32 age
        )
    {
        if ( age > 0
             && age < 50
             && ( age < 20
                  || age > 40 ) )
        {
            Age
                += ( age
                     + ( age - 2 )
                     + ( age
                         * ( age + 10 )
                         * ( age + 20 ) ) )
                   + GetAgeOffset(
                         age * 2
                         - 20
                         );
        }
        else if ( age > 20
                  && age < 40
                  && ( age < 25
                       || age > 35 ) )
        {
            Age = Math::Round( (double)age * 0.5 );
        }
        else
        {
            Age = age + 10;
        }
    }
};

// -- FUNCTIONS

Dictionary<String, int32>* GetAgeInterval(
    const Array<Person>& sortedPersonArray
    )
{
    if ( sortedPersonArray.Count() == 0 )
    {
        return nullptr;
    }
    else
    {
        auto ageIntervalDictionary = new Dictionary<String, int32>();

        ageIntervalDictionary->Add(
            TEXT( "firstAge" ),
            sortedPersonArray[ 0 ].Age
            );

        ageIntervalDictionary->Add(
            TEXT( "lastAge" ),
            sortedPersonArray[ sortedPersonArray.Count() - 1 ].Age
            );

        return ageIntervalDictionary;
    }
}

// ~~

void Main(
    )
{
    int32 passIndex = 0;

    while ( passIndex < 5 )
    {
        ++passIndex;
    }

    do
    {
        ++passIndex;
    }
    while ( passIndex < 10 );

    Array<Person> personArray;

    personArray.Add( Person( TEXT( "Mike" ), 49, 85, 1 ) );
    personArray.Add( Person( TEXT( "Luke" ), 30, 77, 0 ) );
    personArray.Add( Person( TEXT( "John" ), 30, 72, 3 ) );

    personArray.Sort(
        []( const Person& firstPerson, const Person& secondPerson )
        {
            try
            {
                if ( firstPerson.Age != secondPerson.Age )
                {
                    return firstPerson.Age - secondPerson.Age;
                }
                else
                {
                    return (int32)Math::Sign( firstPerson.Weight - secondPerson.Weight );
                }
            }
            catch ( const Exception& error )
            {
                LOG( Error, error.ToString() );
            }

            return 0;
        }
        );

    auto ageIntervalDictionary = GetAgeInterval( personArray );

    if ( ageIntervalDictionary != nullptr )
    {
        LOG(
            Info,
            TEXT( "First age: {0}" ),
            ageIntervalDictionary->At( TEXT( "firstAge" ) )
            );

        LOG(
            Info,
            TEXT( "Last age: {0}" ),
            ageIntervalDictionary->At( TEXT( "lastAge" ) )
            );
    }
    else
    {
        LOG( Info, TEXT( "No age interval" ) );
    }
}

File: IN/sample.js
// -- CONSTANTS

const
    minimumPassCount = 0,
    maximumPassCount = 5;

// -- TYPES

class Being
{
    // -- CONSTRUCTORS

    constructor(
        name,
        age
        )
    {
        this.name = name;
        this.age = age;
    }
}

// ~~

class Person
    extends Being
{
    // -- CONSTRUCTORS

    constructor(
        name,
        age,
        weight,
        dogCount
        )
    {
        super( name, age );
        this.weight = weight;
        this.dogCount = dogCount;
    }

    // -- INQUIRIES

    getAge(
        )
    {
        return this.age;
    }

    // ~~

    getAgeOffset(
        otherAge
        )
    {
        return otherAge - this.age;
    }

    // ~~

    getHelloMessage(
        )
    {
        return `Hello, my name is ${ this.name }, I'm ${ this.age } years old and I weight ${ this.weight } kilograms.`;
    }

    // -- OPERATIONS

    setAge(
        age
        )
    {
        this.age = age;
    }

    // ~~

    setFakeAge(
        age
        )
    {
        if ( age > 0
             && age < 50
             && ( age < 20
                  || age > 40 ) )
        {
            this.age
                += ( age
                     + ( age - 2 )
                     + ( age
                         * ( age + 10 )
                         * ( age + 20 ) ) )
                   + this.getAgeOffset(
                         age * 2
                         - 20
                         );
        }
        else if ( age > 20
                  && age < 40
                  && ( age < 25
                       || age > 35 ) )
        {
            this.age = Math.round( age * 0.5 );
        }
        else
        {
            this.age = age + 10;
        }
    }
}

// -- FUNCTIONS

function getAgeInterval(
    sortedPersonArray
    )
{
    if ( sortedPersonArray.length === 0 )
    {
        return null;
    }
    else
    {
        return (
            {
                firstAge: sortedPersonArray[ 0 ].age,
                lastAge: sortedPersonArray[ sortedPersonArray.length - 1 ].age
            }
            );
    }
}

// ~~

function main(
    )
{
    let passIndex = 0;

    while ( passIndex < 5 )
    {
        ++passIndex;
    }

    do
    {
        ++passIndex;
    }
    while ( passIndex < 10 );

    let personArray
        = [
            new Person( 'Mike', 49, 85, 1 ),
            new Person( 'Luke', 30, 77, 0 ),
            new Person( 'John', 30, 72, 3 )
        ];

    personArray.sort(
        ( firstPerson, secondPerson ) =>
        {
            try
            {
                if ( firstPerson.age !== secondPerson.age )
                {
                    return firstPerson.age - secondPerson.age;
                }
                else
                {
                    return firstPerson.weight - secondPerson.weight;
                }
            }
            catch ( error )
            {
                console.log( error.toString() );
            }

            return 0;
        }
        );

    let ageInterval = getAgeInterval( personArray );

    if ( ageInterval !== null )
    {
        console.log( `First age: ${ ageInterval.firstAge }` );
        console.log( `Last age: ${ ageInterval.lastAge }` );
    }
    else
    {
        console.log( 'No age interval' );
    }
}

// -- STATEMENTS

main();
